File: models/transaction.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type TransactionStatus string

const (
	TransactionStatusPending   TransactionStatus = "pending"
	TransactionStatusConfirmed TransactionStatus = "confirmed"
	TransactionStatusFailed    TransactionStatus = "failed"
	TransactionStatusInvalid   TransactionStatus = "invalid"
)

type Transaction struct {
	ID            *primitive.ObjectID `json:"_id,omitempty" bson:"_id,omitempty"`
	Hash          string              `json:"hash" bson:"hash"`
	FromAddress   string              `json:"from_address" bson:"from_address"`
	ToAddress     string              `json:"to_address" bson:"to_address"`
	BlockHeight   uint64              `json:"block_height" bson:"block_height"`
	Confirmations uint64              `json:"confirmations" bson:"confirmations"`
	Chain         Chain               `bson:"chain" json:"chain"`
	Status        TransactionStatus   `json:"status" bson:"status"`
	CreatedAt     time.Time           `bson:"created_at" json:"created_at"`
	UpdatedAt     time.Time           `bson:"updated_at" json:"updated_at"`
	Refund        *primitive.ObjectID `json:"refund" bson:"refund"`
	Message       *primitive.ObjectID `json:"message" bson:"message"`
}

type MintMemo struct {
	Address string `json:"address" bson:"address"`
	ChainID string `json:"chain_id" bson:"chain_id"`
}

File: e2e/src/types/transaction.ts
// Import the required dependencies for primitive types and date handling
import { ObjectId, Long } from 'mongodb';
import { Hex } from 'viem';

// Assuming Chain type is imported from a local file
import { Chain } from './node';

// Define the TransactionStatus type
export type TransactionStatus = 'pending' | 'confirmed' | 'failed' | 'invalid';

// Define the Transaction type
export type Transaction = {
  readonly _id: ObjectId;
  readonly hash: Hex;
  readonly from_address: Hex;
  readonly to_address: Hex;
  readonly block_height: Long;
  readonly confirmations: Long;
  readonly chain: Chain;
  readonly status: TransactionStatus;
  readonly created_at: Date;
  readonly updated_at: Date;
  readonly refund?: ObjectId | null;
  readonly messages: ObjectId[];
};

// Define the MintMemo type
export type MintMemo = {
  readonly address: Hex;
  readonly chain_id: string;
};

export const CollectionTransactions = 'transactions';
